[mlir][SVE] Add an e2e test for vector.contract

banach-space opened the pull request:
Adds an end-to-end test for `vector.contract` that targets SVE (i.e.
scalable vectors). Note that this requires lifting the restriction on
`vector.outerproduct` (to which `vector.contract` is lowered to) that
would deem the following as invalid by the Op verifier (*):

```
vector.outerproduct %27, %28, %26 {kind = #vector.kind<add>} : vector<3xf32>, vector<[2]xf32>
```

This is indeed valid as the end-to-end test demonstrates (at least when
compiling for SVE).

banach-space committed:
[mlir][SVE] Add an e2e test for vector.contract

Adds an end-to-end test for `vector.contract` that targets SVE (i.e.
scalable vectors). Note that this requires lifting the restriction on
`vector.outerproduct` (to which `vector.contract` is lowered to) that
would deem the following as invalid by the Op verifier (*):

```
vector.outerproduct %27, %28, %26 {kind = #vector.kind<add>} : vector<3xf32>, vector<[2]xf32>
```

This is indeed valid as the end-to-end test demonstrates (at least when
compiling for SVE).

Depends on #68794

dcaballe commented:
Is this for review?

banach-space commented:
> Is this for review?

Yes, but only the last commit - first 2 are just dependencies. 

Apologies for not updating the summary earlier (the default generated by GitHub when uploading multiple commits is always confusing).

MacDue commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSVE/test-contraction.mlir:
```diff
@@ -0,0 +1,137 @@
+  %0 = vector.mask %m { vector.contract #matmat_trait %vector_a, %vector_b, %vector_c
+    : vector<3x5xi32>, vector<5x[2]xi32> into vector<3x[2]xi32> } : vector<3x[2]x5xi1> -> vector<3x[2]xi32>
+
+  // Print the output
+  %slice1 = vector.extract %0[0] : vector<[2]xi32> from vector<3x[2]xi32>
+  // I32: ( 75959, 75959, 75959, 75959
+  vector.print %slice1 : vector<[2]xi32>
+  %slice2 = vector.extract %0[1] : vector<[2]xi32> from vector<3x[2]xi32>
+  // I32-NEXT: ( 75959, 75959, 75959, 75959
+  vector.print %slice2 : vector<[2]xi32>
+  %slice3 = vector.extract %0[2] : vector<[2]xi32> from vector<3x[2]xi32>
+  // I32-NEXT: ( 75959, 75959, 75959, 75959
+  vector.print %slice3 : vector<[2]xi32>
```
To be safe this should only test the base size (i.e. the first two elements here):
```suggestion
  // Print the output
  %slice1 = vector.extract %0[0] : vector<[2]xi32> from vector<3x[2]xi32>
  // I32: ( 75959, 75959
  vector.print %slice1 : vector<[2]xi32>
  %slice2 = vector.extract %0[1] : vector<[2]xi32> from vector<3x[2]xi32>
  // I32-NEXT: ( 75959, 75959
  vector.print %slice2 : vector<[2]xi32>
  %slice3 = vector.extract %0[2] : vector<[2]xi32> from vector<3x[2]xi32>
  // I32-NEXT: ( 75959, 75959
  vector.print %slice3 : vector<[2]xi32>
```

banach-space replied:
> To be safe this should only test the base size

This is not the first time I'm doing this mistake today 😂 Thanks!

MacDue commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSVE/test-contraction.mlir:
```diff
@@ -0,0 +1,137 @@
+    : vector<3x5xf32>, vector<5x[2]xf32> into vector<3x[2]xf32> } : vector<3x[2]x5xi1> -> vector<3x[2]xf32>
+
+  // Print the output
+  %slice1 = vector.extract %0[0] : vector<[2]xf32> from vector<3x[2]xf32>
+  // F32: ( 10.7045, 10.7045, 10.7045, 10.7045
+  vector.print %slice1 : vector<[2]xf32>
+  %slice2 = vector.extract %0[1] : vector<[2]xf32> from vector<3x[2]xf32>
+  // F32-NEXT: ( 10.7045, 10.7045, 10.7045, 10.7045
+  vector.print %slice2 : vector<[2]xf32>
+  %slice3 = vector.extract %0[2] : vector<[2]xf32> from vector<3x[2]xf32>
+  // F32-NEXT: ( 10.7045, 10.7045, 10.7045, 10.7045
+  vector.print %slice3 : vector<[2]xf32>
```
Ditto here :)

MacDue commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSVE/test-contraction.mlir:
```diff
@@ -0,0 +1,137 @@
+  %i32_123 = arith.constant 123 : i32
+  %i32_314 = arith.constant 314 : i32
+
+  // Allocate and initialize matrix A
+  %A_alloc = memref.alloca() : memref<3x5xi32>
+  linalg.fill ins(%i32_123 : i32) outs(%A_alloc :memref<3x5xi32>)
+  %mask_a = vector.create_mask %c3, %c5 : vector<3x5xi1>
+  %vector_a = vector.transfer_read %A_alloc[%c0, %c0], %cst, %mask_a {in_bounds = [true, true]} : memref<3x5xi32>, vector<3x5xi32>
```
Are these memrefs/loads needed?
```suggestion
  // Setup vector A:
  %vector_a = arith.constant : dense<123> : vector<3x5xi32>
```
(same comment elsewhere)

banach-space replied:
Nope, I was debugging something weird and just wanted to avoid the bufferisation pass. But that's has been resolved now :)

banach-space committed:
fixup! [mlir][SVE] Add an e2e test for vector.contract

Simplfiy following Ben's advice

MacDue commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSVE/test-contraction.mlir:
```diff
@@ -0,0 +1,116 @@
+  %c5 = arith.constant 5 : index
+  %n_rows = arith.muli %vscale, %c2 : index
+
+  %cst = arith.constant 0: i32
```
```suggestion
```
This is unused.

MacDue commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSVE/test-contraction.mlir:
```diff
@@ -0,0 +1,116 @@
+  %c5 = arith.constant 5 : index
+  %n_rows = arith.muli %vscale, %c2 : index
+
+  %cst = arith.constant 0.0: f32
+  %f32_123 = arith.constant 1.23 : f32
+  %f32_314 = arith.constant 3.14 : f32
```
```suggestion
```
These are unused.

MacDue commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSVE/test-contraction.mlir:
```diff
@@ -0,0 +1,116 @@
+  %vector_c = arith.constant dense<3.14> : vector<3x[2]xf32>
+
+  // Matmul
+  %m = vector.create_mask %c3, %n_rows, %c5 : vector<3x[2]x5xi1>
```
What's the point of the masking here? This could be:
```suggestion
  %m = arith.constant dense<true> : vector<3x[2]x5xi1>
```
Which would eliminate the other extra constants (same for the other test). But I'm not sure the mask is needed since it's all true (and will likely get folded away).

banach-space committed:
fixup! [mlir][SVE] Add an e2e test for vector.contract

Remove masks and unused variables, make sure that the file is compiled only once, add comments

MacDue reviewed (approved):
LGTM! Thanks for the changes :)
